docs/dev/convert_layout.rst

```diff
@@ -0,0 +1,192 @@
+        kernel_layout = attrs['kernel_layout']
+        data, weight = inputs
+        assert desired_layout == 'NCHW', \
+                "Currently only transformation to NCHW layout is supported."
+        if desired_layout == 'NCHW':
+            new_attrs = dict(attrs)
+            new_attrs['data_layout'] = desired_layout
+            new_attrs['kernel_layout'] = 'OIHW'
+
+            if data_layout == 'NHWC' and kernel_layout == 'HWIO':
+                # Convert (NHWC, HWIO) to (NCHW, OIHW)
+                return relay.nn.conv2d(data, weight, **new_attrs)
+            if data_layout == 'NHWC' and kernel_layout == 'HWOI':
+                # Convert (NHWC, HWOI) to (NCHW, OIHW). Depthwise conv2d.
+                return relay.nn.conv2d(data, weight, **new_attrs)
+        return None
+
+
+**Layout inference** - Relay op has an attribute - *FInferCorrectLayout* - that developers can implement to handle data layouts. Currently, this attribute is only exposed in C++. This function takes original input layouts and the new input layouts (passed from the previous operator or from the python callback for layout alteration). A TVM developer can use this function to infer the final data layout and also modify the op attributes if needed.
+
+This component is used for *lightly-layout sensitive* operators. We try to accept the new input layout, and modify the current operator attributes (like axis for concatenate, pad_width for pad) to adapt to the new data layout. By accepting the new input data layout, we prevent the insertion of a layout transform. In absence of this function, Layout rewrite might have to insert a layout transform, if the previous operator has a different output data layout than the original one. One example to adapt to NCHW data layout is presented here for Batch Norm operator.
+
+.. code-block:: c++
+
+    Array<Array<Layout>> BatchNormInferCorrectLayout(const Attrs& attrs,
+                                                     const Array<Layout>& new_in_layouts,
+                                                     const Array<Layout>& old_in_layouts,
+                                                     const Array<Array<IndexExpr>>& old_in_shapes) {
+      BatchNormAttrs* param = const_cast<BatchNormAttrs*>(attrs.as<BatchNormAttrs>());
+
+      size_t axis =
+          param->axis < 0 ? param->axis + old_in_shapes[0].size() : static_cast<size_t>(param->axis);
+
+      Layout ret = Layout::Undef();
+
+      // If new_in_layouts are defined, this code tries to modify the layout.
+      if (new_in_layouts.defined() && old_in_layouts.defined()) {
+        // Get the new C axis. Extract the dim in old layout. Find the index of that dim in next layout.
+        const auto& bn_dim = old_in_layouts[0][axis];
+        auto new_index = new_in_layouts[0].IndexOf(bn_dim);
+        param->axis = new_index;
+        ret = new_in_layouts[0];
+      } else if (old_in_layouts.defined()) {
+        ret = old_in_layouts[0];
+      }
```

Comment:
If `old_in_layouts` isn't defined, ret is just left undefined. Will that cause an error or does it somehow work itself out?